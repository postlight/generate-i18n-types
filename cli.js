#!/usr/bin/env node

const argv = process.argv.slice(2);
const fs = require('fs');
const yaml = require('js-yaml');
const path = require('path');
const { flattenDeep } = require('lodash');

const { translation } = yaml.safeLoad(
  fs.readFileSync(path.resolve(__dirname, argv.translations), 'utf8')
);

export const getKeyStrings = (obj, initialKey = '') => {
  const keys = Reflect.ownKeys(obj);
  return keys.map(key =>
    typeof obj[key] === 'string'
      ? `${initialKey}.${key}`
      : getKeyStrings(obj[key], `${initialKey}.${key}`)
  );
};

export const getString = (obj, key) =>
  key.split('.').reduce((acc, key) => acc[key], obj);

export const getAllStrings = obj => {
  return flattenDeep(getKeyStrings(obj)).map(s => s.slice(1));
};

export const toType = key =>
  key
    .split('.')
    .map(s => s.slice(0, 1).toUpperCase() + s.slice(1))
    .join('');

export const generateTypes = async obj => {
  // Turn json to key strings
  const keyStringsArr = getAllStrings(obj);

  // Generate string enum names that map to key strings
  const stringsAsEnumProp = keyStringsArr.map(toType);

  // Identify strings that require data
  const allStrings = keyStringsArr.map(key => getString(obj, key));
  const dataRe = /\{\{(\w+)\}\}/g;
  const stringsWithData = allStrings.map(str => {
    let match = dataRe.exec(str);
    const matches = [];
    while (match) {
      matches.push(match[1]);
      match = dataRe.exec(str);
    }
    return matches.length > 0 ? matches : null;
  });

  // Generate the enum
  const noArgsStrings = keyStringsArr.filter(
    (k, index) => !stringsWithData[index]
  );
  const stringsWithArgs = keyStringsArr.filter(
    (k, index) => stringsWithData[index]
  );

  const stringArgsEnum = 'I18NStringsWithArgs';
  const stringEnum = 'I18NStrings';
  const comments = `/* tslint:disable */\n/* eslint-disable */\n// This file was automatically generated and should not be edited.`;

  const enumsWithArgsString = `${stringsWithArgs.reduce((acc, key, index) => {
    return `${acc}
  ${toType(key)} = '${key}',`;
  }, `export enum ${stringArgsEnum} {`)}
}`;
  const enumNoArgsString = `${noArgsStrings.reduce((acc, key, index) => {
    return `${acc}
  ${toType(key)} = '${key}',`;
  }, `export enum ${stringEnum} {`)}
}`;

  // Write enums
  fs.writeFileSync(
    `./${argv.stringTypes}`,
    `${comments}\n\n${enumsWithArgsString}\n\n${enumNoArgsString}`
  );

  // Write overloaded signatures with data requirements
  const overloadedSignature = `function translate(
  key: ${stringArgsEnum}.REPLACE_ENUM,
  REPLACE_DATA
): string;`;
  const overloadedSignatures = `${stringsWithData
    .map((arr, index) => {
      if (arr === null) {
        return;
      }
      return overloadedSignature
        .replace('REPLACE_ENUM', stringsAsEnumProp[index])
        .replace(
          'REPLACE_DATA',
          `{ ${arr.join(', ')} }: { ${arr
            .map(key => `${key}: string`)
            .join(';')} }`
        );
    })
    .filter(Boolean)
    .join('\n')}\nfunction translate(key: ${stringEnum}): string;`;

  // Write content to util file
  const i18nPath = path.resolve(__dirname, argv.util);
  const i18nPathDir = argv.util.slice(0, argv.util.lastIndexOf('/'));
  const stringTypesPath = argv.stringTypes.replace(
    path.extname(argv.stringTypes),
    ''
  );

  const readTranslation =
    'const en = yaml.safeLoad(\n' +
    `    fs.readFileSync('${path.relative(
      i18nPathDir,
      argv.translations
    )}', 'utf8')\n` +
    `);\n\n`;

  const sigils = [
    '// ==== START OVERLOADED SIGNATURES',
    '// ==== END OVERLOADED SIGNATURES',
  ];

  const imports = `import i18n from 'i18next';\nimport yaml from 'js-yaml';\nimport fs from 'fs';\nimport { ${stringEnum}, ${stringArgsEnum} } from '${path.relative(
    i18nPathDir,
    stringTypesPath
  )}';`;

  const translate =
    'function translate(key: string, data?: { [key: string]: string }) {\n' +
    '    if (boundT) {\n' +
    '        return boundT(key, data);\n' +
    '    }\n' +
    '    i18n.init({\n' +
    "        lng: 'en',\n" +
    '        resources: { en },\n' +
    '        interpolation: {\n' +
    '            escapeValue: false,\n' +
    '        },\n' +
    '    });\n' +
    '    const { t } = i18n;\n' +
    '    boundT = t.bind(i18n);\n' +
    '    return boundT(key, data);\n' +
    '}\n\nexport default translate;';

  const content = `${comments}\n\n${imports}\n\nexport {  ${stringEnum}, ${stringArgsEnum} } from '${path.relative(
    i18nPathDir,
    stringTypesPath
  )}';\n\nlet boundT: typeof i18n.t;\n\n${readTranslation}${
    sigils[0]
  }\n${overloadedSignatures}\n${sigils[1]}\n\n${translate}`;

  fs.writeFileSync(i18nPath, content);
};

generateTypes(translation);

module.exports = {
  getString,
  getKeyStrings,
  getAllStrings,
  generateTypes,
  toType,
};
