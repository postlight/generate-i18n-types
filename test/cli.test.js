import fs from 'fs';
import { exec } from 'child_process';
import { generateTypes } from '../cli';
import {
  toType,
  getAllStrings,
  getKeyStrings,
  getString,
} from '../cli-helpers';
import translation from './test-helpers';

jest.mock('fs');

const argv = [...process.argv];

const cli = args => {
  return new Promise(resolve => {
    exec(`node cli.js ${args.join(' ')} --output`, (error, stdout, stderr) => {
      resolve({
        error,
        stdout,
        stderr,
      });
    });
  });
};

describe('generate-i18n-types', () => {
  it('should have toType return correct type format', () => {
    const key = 'authFirst.text';
    expect(toType(key)).toBe('AuthFirstText');
  });

  it('should have getAllStrings return array of keys', () => {
    const stringsArray = [
      'appName',
      'supportEmail',
      'authFirst.text',
      'errors.invalidUsers',
    ];
    expect(getAllStrings(translation)).toStrictEqual(stringsArray);
  });

  it('should have getString return string for key', () => {
    const key = 'authFirst.text';
    expect(getString(translation, key)).toBe(
      'Before we can create your account, please <{{url}}|authenticate with My App>.'
    );
  });

  it('should have getKeyStrings return array of initial keys', () => {
    const keyStringsArray = [
      '.appName',
      '.supportEmail',
      ['.authFirst.text'],
      ['.errors.invalidUsers'],
    ];
    expect(getKeyStrings(translation)).toStrictEqual(keyStringsArray);
  });

  it('should have generateTypes generate 2 files', async () => {
    await generateTypes(translation);
    expect(fs.writeFileSync).toBeCalledTimes(2);
    expect(fs.writeFileSync.mock.calls).toMatchSnapshot();
  });

  it('should return error message if missing a required argument', async () => {
    argv.pop();
    const result = await cli(argv);
    expect(result.stdout).toContain('Error: Missing required arguments');
  });
});
