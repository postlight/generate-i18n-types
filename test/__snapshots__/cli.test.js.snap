// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate-i18n-types should have generateTypes generate 2 files 1`] = `
Array [
  Array [
    "./src/types/strings.ts",
    "/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

export enum I18NStringsWithArgs {
  AuthFirstText = 'authFirst.text',
  ErrorsInvalidUsers = 'errors.invalidUsers',
}

export enum I18NStrings {
  AppName = 'appName',
  SupportEmail = 'supportEmail',
}",
  ],
  Array [
    "./src/utils/i18n.ts",
    "/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

import i18n from 'i18next';
import yaml from 'js-yaml';
import fs from 'fs';
import { I18NStrings, I18NStringsWithArgs } from '../types/strings';

export {  I18NStrings, I18NStringsWithArgs } from '../types/strings';

let boundT: typeof i18n.t;

const en = yaml.safeLoad(
    fs.readFileSync('../../locales/en/translation.yml', 'utf8')
);

// ==== START OVERLOADED SIGNATURES
function translate(
  key: I18NStringsWithArgs.AuthFirstText,
  { url }: { url: string }
): string;
function translate(
  key: I18NStringsWithArgs.ErrorsInvalidUsers,
  { invalidUsers }: { invalidUsers: string }
): string;
function translate(key: I18NStrings): string;
// ==== END OVERLOADED SIGNATURES

function translate(key: string, data?: { [key: string]: string }) {
    if (boundT) {
        return boundT(key, data);
    }
    i18n.init({
        lng: 'en',
        resources: { en },
        interpolation: {
            escapeValue: false,
        },
    });
    const { t } = i18n;
    boundT = t.bind(i18n);
    return boundT(key, data);
}

export default translate;",
  ],
]
`;
